### ISSUE TITLE
Suspicious state from checker verilog-verilator.

### ISSUE THREAD

**Issue by AngelTerrones** [flycheck/flycheck#740]
Hi. I got the following error, trying to lint a simple verilog module (or, any verilog file):
```
Suspicious state from syntax checker verilog-verilator: Checker verilog-verilator returned non-zero exit code 126, but no errors from output: emacs: /usr/local/bin/verilator: Exec format error

Checker definition probably flawed.
```
I checked  using `flycheck-compile`, and I didn't get any error.

My emacs:
```
GNU Emacs 24.5.1 (x86_64-unknown-linux-gnu, GTK+ Version 3.16.2) of 2015-04-20 on bitzer.hoetzel.info
```

Flycheck:
```
Flycheck version: 0.25snapshot (package: 20150821.47)
```

Verilator:
```
Verilator 3.876 2015-08-12 rev verilator_3_876
```

**Comment by lunaryorn** [comment]
@AngelTerrones If `M-x flycheck-compile` works it's probably because it picks a different `verilator` executable due to a different `$PATH` setting (`flycheck-compile` goes through the shell, whereas a normal syntax check does not).   Can you actually run `/usr/local/bin/verilator` from the terminal, i.e. outside Emacs?  Do you have multiple `verilator` installations?

**Comment by AngelTerrones** [comment]
Hi.
```
/usr/local/bin/verilator --lint-only -Wall -I Add.v
```
Returns with exit code 0, no errors, no warnings (the file is ok). The same file inside emacs give the `suspicious...` error.
I have one installation.

**Comment by lunaryorn** [comment]
What is the output of `file /usr/local/bin/verilator`?

**Comment by AngelTerrones** [comment]
```
/usr/local/bin/verilator
Usage:
        verilator --help
        verilator --version
        verilator --cc [options] [top_level.v]... [opt_c_files.cpp/c/cc/a/o/so]
        verilator --sc [options] [top_level.v]... [opt_c_files.cpp/c/cc/a/o/so]
        verilator --lint-only    [top_level.v]...

```

With exit code 2 (using `echo $?`).

**Comment by lunaryorn** [comment]
The output of `head /usr/local/bin/verilator` then?  At a wild guess it's something like `/usr/bin/env perl`, and you don't have a `perl` executable in your Emacs' `exec-path`…

**Comment by AngelTerrones** [comment]
```
: # -*-Mode: perl;-*- use perl, wherever it is
eval 'exec perl -wS $0 ${1+"$@"}'
  if 0;
######################################################################
#
# Copyright 2003-2015 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
#

```

**Comment by AngelTerrones** [comment]
My emacs `exec-path` includes `/usr/bin`, and there is where perl lives.

**Comment by lunaryorn** [comment]
Where's the shebang?!

**Comment by AngelTerrones** [comment]
Good question /:

Let me try with the shebang.

**Comment by AngelTerrones** [comment]
I removed the first line, and put the shebang. It works now, so I will close this, and report to verilator repo.

Thanks for all.

**Comment by ackalker** [comment]
@lunaryorn Please reopen this issue, it is not invalid. Firstly, the script is actually a shell script with Perl code embedded in it. The reason why it doesn't simply use a shebang is because many (most?) image loaders don't support passing more than one argument to the interpreter, which this script must do. The weird-looking first line is an established practice to avoid the script being run by `/bin/csh` rather than `bin/sh`, see: [Why do some scripts start with #! ... ?](http://www.faqs.org/faqs/unix-faq/faq/part3/section-16.html) from the Unix FAQ for an explanation why it is done this way.

**Comment by ackalker** [comment]
Workaround for users who have a "user" bin directory in their `PATH`: create a file `.../bin/verilator` with contents:
```sh
#!/bin/sh
exec /path/to/verilator "$@"
```
and make it executable using `chmod +x .../bin/verilator`

**Comment by lunaryorn** [comment]
@ackalker What you've written and linked to explains why the script looks the way it looks but at least to me it doesn't provide a clue on what's wrong in this specific case, and beyond historical reference I'm not even sure whether the article you linked to is still relevant.  It's over 20 years old, after all.

I can well reopen this issue but I'd like to be sure whether this is actually an issue in Flycheck.  In other words, what are we supposed to do to address this problem?

**Comment by lunaryorn** [comment]
Besides even considering information from the article I'm still not sure why the script doesn't carry a /bin/sh shebang.

**Comment by PrgrmAtCeleritas** [comment]
I was expereincing this same issue and @ackalker's workaround worked for me

**Comment by gmcastil** [comment]
I am also having this problem, and the workaround that @ackalker suggested no longer works.  Has anyone figured out what the actual problem is?

**Comment by CeleritasCelery** [comment]
@gmcastil what happens when you run your sh script?

**Comment by gmcastil** [comment]
I was able to fix it, but needed to add the following to my ``.emacs``.
```elisp
(setq flycheck-verilog-verilator-executable "invoke-verilator.sh"
```
This is the ``invoke-verilator.sh`` that I have in my path.
```bash
#!/bin/bash
exec $HOME/.local/bin/verilator "$@"
```
Verilator is actually installed in two places on this machine - an older version in ``/usr/bin`` and the latest version I built myself installed in my home directory.  Hope this helps someone.
